cpp/algo_module/algorithm/algo.cpp

```diff
@@ -1 +1,204 @@
+    return true;
+  }
+
+  if (in && config.in_rels.find(std::string(rel.Type())) != config.in_rels.end()) {
+    return true;
+  }
+
+  if (!in && config.out_rels.find(std::string(rel.Type())) != config.out_rels.end()) {
+    return true;
+  }
+
+  return false;
+}
+
+bool Algo::LabelOk(const mgp::Node &node, const Config &config) {
+  bool whitelist_empty = config.whitelist.empty();
+
+  for (auto label : node.Labels()) {
+    if (config.blacklist.find(std::string(label)) != config.blacklist.end()) {
+      return false;
+    }
+    if (!whitelist_empty && config.whitelist.find(std::string(label)) == config.whitelist.end()) {
+      return false;
+    }
+  }
+  return true;
+}
+void Algo::ParseRelationships(const std::shared_ptr<NodeObject> &prev, bool in, const GoalNodes &nodes, Lists &lists,
+                              const Config &config) {
+  auto rels = in ? prev->node.InRelationships() : prev->node.OutRelationships();
+  for (const auto rel : rels) {
+    if (!RelOk(rel, config, in)) {
+      continue;
+    }
+    const auto node = in ? rel.From() : rel.To();
+    if (!LabelOk(node, config)) {
+      continue;
+    }
+    auto heuristic = CalculateHeuristic(config, node, nodes) * config.epsilon;  // epsilon 0 == UCS
+    auto distance = CalculateDistance(config, rel);
+    std::shared_ptr<NodeObject> nb =
+        std::make_shared<NodeObject>(heuristic, distance + prev->total_distance, node, rel, prev);
+    if (!lists.closed.FindAndCompare(nb)) {
+      continue;
+    }
+    lists.open.Insert(nb);
+  }
+}
+
+std::shared_ptr<Algo::NodeObject> Algo::InitializeStart(const mgp::Node &start) {
+  if (start.InDegree() == 0 && start.OutDegree() == 0) {
+    throw mgp::ValueException("Start node must have in or out relationships!");
+  }
+
+  if (start.InDegree() != 0) {
+    return std::make_shared<NodeObject>(0, 0, start, *start.InRelationships().begin(), nullptr);
+  }
+
+  return std::make_shared<NodeObject>(0, 0, start, *start.OutRelationships().begin(), nullptr);
+}
+
+std::pair<mgp::Path, double> Algo::HelperAstar(const GoalNodes &nodes, const Config &config) {
+  Lists lists = Lists();
+
+  std::shared_ptr<NodeObject> start_nb = InitializeStart(nodes.start);
+  lists.open.Insert(start_nb);
+
+  while (!lists.open.Empty()) {
+    auto nb = lists.open.Top();
+    lists.open.Pop();
+    if (nb->node == nodes.target) {
+      return BuildResult(nb, nodes.start);
+    }
+    lists.closed.Insert(nb);
+    ParseRelationships(nb, false, nodes, lists, config);
+    ParseRelationships(nb, true, nodes, lists, config);
+  }
+  return std::pair<mgp::Path, double>(mgp::Path(nodes.start), 0);
+}
+
+std::pair<mgp::Path, double> Algo::BuildResult(std::shared_ptr<NodeObject> final, const mgp::Node &start) {
+  mgp::Path path = mgp::Path(start);
+  std::vector<mgp::Relationship> rels;
+
+  double weight = final->total_distance;
+  while (final->prev) {
+    rels.push_back(final->rel);
+    final = final->prev;
```

Comment:
final is keyword in c++, usually used for classes. Can you change name of this one to final_node or something like that